aws-java-sdk-core/src/main/java/com/amazonaws/auth/InstanceProfileCredentialsProvider.java

```diff
@@ -85,7 +102,7 @@ public InstanceProfileCredentialsProvider(boolean refreshCredentialsAsync) {
                 @Override
                 public void run() {
                     try {
-                        credentialsFetcher.getCredentials();
+                        if (eagerlyRefreshCredentialsAsync || getCredentialsCalled) credentialsFetcher.getCredentials();
```

Comment:
I don't think this does what is intended. The goal is to be safe (as in, not fail every invocation of `run()` when not on an EC2). Given `new InstanceProfileCredentialsProvider(true)` or `new InstanceProfileCredentialsProvider(true, true)` does not do that because it will attempt to refresh. The only safe option for EC2 and non-EC2 is `new InstanceProfileCredentialsProvider(true, false)` so that should really be the only option (i.e. remove `eagerlyRefreshCredentialsAsync`).